simulator/src/main/java/com/hazelcast/simulator/agent/CommandFuture.java

```diff
@@ -58,32 +58,33 @@ public E get() throws InterruptedException, ExecutionException {
             if (result instanceof Throwable) {
                 throw new ExecutionException((Throwable) result);
             }
-            return (E) result;
+            return result;
         }
     }
 
     @Override
-    public E get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
-        long remainingTimeoutMs = unit.toMillis(timeout);
+    public E get(long timeout, TimeUnit timeUnit) throws InterruptedException, ExecutionException, TimeoutException {
+        if (timeout < 0 || timeUnit == null) {
+            throw new IllegalArgumentException("Invalid timeout or timeUnit for CommandFuture.get()");
+        }
 
+        long remainingTimeoutNanos = timeUnit.toNanos(timeout);
         synchronized (this) {
-            for (; ; ) {
-                if (result != NO_RESULT) {
-                    if (result instanceof Throwable) {
-                        throw new ExecutionException((Throwable) result);
-                    }
-                    return (E) result;
-                }
-
-                if (remainingTimeoutMs <= 0) {
-                    throw new TimeoutException("Timeout while executing : "
-                            + command + " total timeout: " + unit.toMillis(timeout) + " ms");
-                }
-
-                long startMs = System.currentTimeMillis();
-                wait(remainingTimeoutMs);
-                remainingTimeoutMs -= System.currentTimeMillis() - startMs;
+            while (result == NO_RESULT && remainingTimeoutNanos > 0) {
```

Comment:
I rewrote this to match the structure of the `get()` method. So this method has now the same logic flow, just with the timeout handling added to it. I also think a `while(condition)` is easier to read than a `while(true)` with the condition hidden within the loop.